system-tests/lib/cre/types.go

```diff
@@ -527,13 +517,259 @@ type DonMetadata struct {
 }
 
+func (m *DonMetadata) IsGateway() bool {
+	return m.ns.GatewayNodeIndex != -1 // don't use flag here b/c may not be set
+}
+
+func (m *DonMetadata) ContainsBootstrapNode() bool {
+	return m.ns.BootstrapNodeIndex != -1 // don't use flag here b/c may not be set
+}
+
+func (m *DonMetadata) RequiresGateway() bool {
+	return slices.Contains(m.Flags, CustomComputeCapability) ||
+		slices.Contains(m.Flags, WebAPITriggerCapability) ||
+		slices.Contains(m.Flags, WebAPITargetCapability) ||
+		slices.Contains(m.Flags, VaultCapability) ||
+		slices.Contains(m.Flags, HTTPActionCapability) ||
+		slices.Contains(m.Flags, HTTPTriggerCapability)
+}
+
+func (m *DonMetadata) RequiresWebAPI() bool {
+	return slices.Contains(m.Flags, CustomComputeCapability) ||
+		slices.Contains(m.Flags, WebAPITriggerCapability) ||
+		slices.Contains(m.Flags, WebAPITargetCapability)
+}
+
+func (m *DonMetadata) IsWorkflowDON() bool {
+	// TODO enum type for DON types
+	return slices.Contains(m.ns.DONTypes, WorkflowDON)
+}
+
+type DonsMetadata struct {
+	dons  []*DonMetadata
+	infra infra.Provider
+}
+
+func NewDonsMetadata(dons []*DonMetadata, infra infra.Provider) (*DonsMetadata, error) {
+	if dons == nil {
+		dons = make([]*DonMetadata, 0)
+	}
+	out := &DonsMetadata{
+		dons:  dons,
+		infra: infra,
+	}
+	return out, out.validate()
+}
+
+func (m DonsMetadata) DonCount() int {
+	return len(m.dons)
+}
+
+func (m DonsMetadata) List() []*DonMetadata {
+	return m.dons
+}
+
+func (m DonsMetadata) validate() error {
+	if len(m.dons) == 0 {
+		return errors.New("at least one don is required")
+	}
+
+	if m.BootstrapNodeCount() == 0 {
+		return errors.New("at least one nodeSet must have a bootstrap node")
+	}
+
+	wfDon, err := m.GetWorkflowDON()
+	if err != nil {
+		return fmt.Errorf("failed to get workflow DON: %w", err)
+	}
+
+	if !wfDon.ContainsBootstrapNode() {
+		return errors.New("due to the limitations of our implementation, workflow DON must always have a bootstrap node")
+	}
+
+	if m.GatewayRequired() && !m.GatewayEnabled() {
+		return errors.New("at least one DON requires gateway due to its capabilities, but no DON is configured with gateway")
+	}
+
+	return nil
+}
+
+func (m DonsMetadata) BootstrapNodeCount() int {
+	count := 0
+	for _, don := range m.dons {
+		if don.ContainsBootstrapNode() {
+			count++
+		}
+	}
+	return count
+}
+
+func (m DonsMetadata) FindByID(id uint64) (*DonMetadata, error) {
+	for _, don := range m.dons {
+		if don.ID == id {
+			return don, nil
+		}
+	}
+	return nil, fmt.Errorf("don with id %d not found", id)
+}
+
+func (m DonsMetadata) FindByName(name string) (*DonMetadata, error) {
+	for _, don := range m.dons {
+		if don.Name == name {
+			return don, nil
+		}
+	}
+	return nil, fmt.Errorf("don with name %s not found", name)
+}
+
+func (m DonsMetadata) FindByFlag(flag string) (*DonsMetadata, error) {
+	dons := make([]*DonMetadata, 0)
+	for _, don := range m.dons {
+		if slices.Contains(don.Flags, flag) {
```

Comment:
I'd suggest using `flags.HasFlag(don.Flags, flag`) since for chain-specific capabilities (e.g. `evm` or `readcontract`) simple contains might not mach anything.